Fix invalid fusion of Matmul and Mul

lgeiger opened the pull request:
This PR fixes #39572 which lead to wrong operator fusion in the TFLite converter for some cases.

This is a resubmission of #40013 which has already be reviewed in detail there.

@rthadur @liufengdb could you take a look?

lgeiger commented:
@liufengdb Could you please take a look?

lgeiger commented:
Any updates on this?

lgeiger commented:
@joker-eph @liufengdb @karimnosseir @gbaned Any news on when this PR can get a review? I think this PR fixes a pretty serious bug in TFLite so it would be great if this could make it into the 2.4 release.

lgeiger commented:
Any updates on this? The main changes of this PR have already been reviewed in #40013 and discussed in #39572 which is a major issue for people using fully connected layers with batch normalisation in TFLite.

gbaned commented:
@liufengdb @karimnosseir Any update on this PR? Please. Thanks!

lgeiger commented:
Any news?

lgeiger commented:
@mihaimaruseac Sorry to ping you here, it looks like this PR is stuck in the pipeline.
It fixes #39572 which has been a serious issue for us since TF 2.2 and is a resubmission of #40013 which has been abandoned but already reviewed in detail.
It would be really good to get this reviewed and merged into 2.4 before the release branch get's cut.

karimnosseir commented:
That got lost for me. I am sorry. I think this pattern was added by @liufengdb, so deferring to him to review.
@lgeiger if got delayed more, ping me and i will check.
Apologies again.

lgeiger commented:
Thanks for the response, looking forward to the review.

lgeiger commented:
@karimnosseir For reference, @liufengdb reviewed the content of this PR in #40013 which wasn't merged since https://github.com/tensorflow/tensorflow/pull/40013#discussion_r452659910 wasn't addressed and the PR abandoned by the original author.

lgeiger commented:
@karimnosseir Any updates?

karimnosseir commented on tensorflow/compiler/mlir/lite/transforms/optimize.cc:
```diff
@@ -372,26 +372,26 @@ struct FuseFullyConnectedAndMul : public OpRewritePattern<TFL::MulOp> {
+    // Check if it is possible to fuse the constant mathematically
+    // output[..., b, j] = (sum_i input[..., b, i] * filter[j, i]) * m[..., b,
+    // j] m[..., b, j] must be independent of ..., b in order to fuse into
+    // filter
```
It will be good if you can expand the comment with more details. By explaining the requirements, why and how it is done.

karimnosseir commented on tensorflow/compiler/mlir/lite/utils/validators.h:
```diff
@@ -70,6 +70,8 @@ inline bool TFPaddingIsSameOrValid(Operation *op, StringAttr *padding) {
 /// Returns whether the given `a` and `b` have broadcast-compatible
 /// types.
 bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b);
+bool IsDimensionsDegenerateExceptLastOne(mlir::Attribute val);
```
Add Function comments please.

karimnosseir commented on tensorflow/lite/python/lite_v2_test.py:
```diff
@@ -1192,5 +1192,81 @@ def model(in_tensor):
+
+      # Check values from converted model.
+      expected_value = root.f(input_data)
+      converter.experimental_new_converter = True
```
You can remove this line - it is the default.

karimnosseir commented on tensorflow/lite/python/lite_v2_test.py:
```diff
@@ -1192,5 +1192,81 @@ def model(in_tensor):
         str(error.exception))
 
 
+class AffineOpThenMulFusionTest(lite_v2_test_util.ModelTest):
```
func1/func2/... are very general names. Please use more descriptive names to what it is testing, and add comments in needed parts please.

Thanks

lgeiger replied:
I refactored the tests to make them clearer and expanded the test coverage.

karimnosseir commented on tensorflow/lite/python/lite_v2_test.py:
```diff
@@ -1192,5 +1192,81 @@ def model(in_tensor):
+      np.testing.assert_almost_equal(expected_value.numpy(), actual_value[0])
+
+      # Enable hybrid quantization, same result
+      converter.experimental_new_converter = True
```
duplicate and not needed. Remove please

lgeiger replied:
Good catch, this code is still from the time when it wasn't enabled by default.

karimnosseir reviewed (changes requested):
Generally LG.
Left few comments.

Thanks

lgeiger committed:
Fix invalid fusion of Matmul and Mul

lgeiger commented:
@karimnosseir Thanks for the review. I rebased the PR onto latest master in b72545ee21dfdc1fde0424a87523536c353f3ce3 and clarified the comments and cleaned up the unittests according the the review suggestions.

Sorry about some of the unclear descriptions, I adapted most of the code from #40013 without too much thinking. I hope this is now clearer to read.

karimnosseir commented on tensorflow/compiler/mlir/lite/transforms/optimize.cc:
```diff
@@ -438,26 +438,26 @@ struct FuseFullyConnectedAndMul : public OpRewritePattern<TFL::MulOp> {
-          rewriter.create<ConstantOp>(mul_op.getLoc(), new_type, new_cst);
-      new_const_val = new_op.getResult();
+    // Only fuse multiplier if all dimensions other than the depth dimension
+    // are equal to 1.
```
Expand comment on why we only fuse in this case.

lgeiger replied:
Thanks for the reply, I tried to improve the comment in 2b2b3a9bd817b05e2451b2de03f170b3282d63f5.

The reason for this is that we can only fuse a multiplier in dimensions that are not summed over during the matmul due to:
![Screenshot 2020-10-20 at 18 07 56](https://user-images.githubusercontent.com/13285808/96620765-c40bcc00-12ff-11eb-9869-37f0b691ed16.png)

lgeiger committed:
Expand explanatory comment

karimnosseir reviewed (approved):
Thanks

lgeiger commented:
@gbaned looks like there are quite a few internal CI failures. Unfortunately I cannot access the logs, is there anything that is still preventing this from being merged from my side?

gbaned commented:
@lgeiger  It is processing internally, we will let you know if anything required from you. Thank you!

lgeiger commented:
@gbaned any updates state of the internal processing?

abattery commented on tensorflow/compiler/mlir/lite/utils/validators.cc:
```diff
@@ -78,5 +78,19 @@ bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b) {
+
+bool IsDimensionsDegenerateExceptLastOne(Attribute val) {
+  if (auto elements = val.dyn_cast<DenseElementsAttr>()) {
+    return IsDimensionsDegenerateExceptLastOne(elements.getType().getShape());
```
This line should be rewritten in order to access the shape information if available.

For example,

if (auto ranked_type = elements.getType().cast<RankedTensorType>()) {
  return IsDimensionsDegenerateExceptLastOne(ranked_type.getShape());
}

Otherwise, it will cause crashing when the type is UnrankedTensorType.

lgeiger replied:
Good catch, I fixed it in 4ebd2633e870dd474fbc9b348d311a4ccdf20158

abattery commented on tensorflow/compiler/mlir/lite/utils/validators.cc:
```diff
@@ -78,5 +78,19 @@ bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b) {
+}
+
+bool IsDimensionsDegenerateExceptLastOne(Attribute val) {
+  if (auto ranked_type = val.getType().cast<RankedTensorType>()) {
```
Sorry for my bad. Let's use dyn_cast to make sure avoid crashing on UnrankedTensorType.

```
if (auto ranked_type = val.getType().dyn_cast<RankedTensorType>()) {
    return IsDimensionsDegenerateExceptLastOne(ranked_type.getShape());
}
```

abattery commented on tensorflow/compiler/mlir/lite/utils/validators.cc:
```diff
@@ -78,5 +78,19 @@ bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b) {
 }
 
+bool IsDimensionsDegenerateExceptLastOne(ArrayRef<int64_t> elements_shape) {
+  for (auto i = 0; i < elements_shape.size() - 1; i++) {
```
zero valued size_t type minus - 1 is problematic and it is causing test failures.

Let's use the ArrayRef member methods as much as possible to avoid unexpected crashing due to the type conversion:

```
  if (elements_shape.empty()) return true;

  for (auto dim : elements_shape.drop_back(1)) {
    if (dim != ShapedType::kDynamicSize) return false;
  }
  return true;
```

abattery replied:
I confused about 1 and -1. Here is the updated suggestion.

```
  if (elements_shape.empty()) return true;

  for (auto dim : elements_shape.drop_back(1)) {
    if (dim != 1) return false;
  }
  return true;
```

lgeiger replied:
Thanks for the review. I addressed the comments in ce4e534ad61012b724b80d1aab2f68017d214e76 however I didn't use `kDynamicSize` is it would change behaviour since we need the dimensions to be `1` and not dynamic, or am I missing something?

abattery commented on tensorflow/compiler/mlir/lite/transforms/optimize.cc:
```diff
@@ -438,26 +438,28 @@ struct FuseFullyConnectedAndMul : public OpRewritePattern<TFL::MulOp> {
+    auto shape = cst.getType().getShape();
+    if (!IsDimensionsDegenerateExceptLastOne(shape)) return failure();
+
+    int64_t element_size = shape.size() > 0 ? shape[shape.size() - 1] : 1;
```
Comparing size() with zero can be easily replaced with empty().

```
    int64_t element_size = shape.empty() ? 1 : shape[shape.size() - 1];
```

lgeiger committed:
Fix shape casting in cases of unranked tensor types

abattery commented:
@lgeiger could you resolve the merge conflict?

abattery commented:
@lgeiger Nevermind. I will handle internal reviews.
